### PR TITLE
Evalcast corrections

### PR DESCRIPTION
**Pull request by dajmcdon**
This should fix both #107 and #112 . 

I added a function argument to `get_predictions()` that implements corrections. For the moment, it's not clear to me that a similar method would work at evaluation time since we won't necessarily have access to most of the data. I put a rather dumb example in the examples (probably shouldn't run at check time).

Also implemented caching via `memoise`. Currently just to memory. The alternatives seem rather diverse and hard to anticipate (at least to me). Also put a unit test to make sure that it worked on two consecutive calls to `get_predictions()` with the same arguments.

Everything passes local checks with some warnings. Most are for undocumented function arguments. Plus all the stupid dplyr notes.

### CODE REVIEW

File: R-packages/evalcast/tests/testthat/test-caching.R [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        signal = c("deaths_incidence_num","confirmed_incidence_num"),
+        start_day=lubridate::ymd("2020-09-15")),
+      lubridate::ymd("2020-10-01"),"epiweek", 1L, "state", "mi"
+    )
+  )
+)
+
+
+test_that("we are caching COVIDCAST calls", {
+  expect_true(t1[1] > t2[1])
```
Comment (jacobbien) [inline R-packages/evalcast/tests/testthat/test-caching.R:31]: This does seem like a reasonable way to test that caching is working.  But I'm wondering if there could be cases where caching occurs but ends up being slower for some reason.  Is there a more direct way to verify that caching occurred? @bnaras, do you have thoughts on this?

File: R-packages/evalcast/data-raw/county_label_lookup.R [hunk 2]
```diff
@@ -0,0 +1,5 @@
+## code to prepare `county_label_lookup` dataset goes here
+county_label_lookup <- paste(
```
Comment (jacobbien) [inline R-packages/evalcast/data-raw/county_label_lookup.R:2]: Where is `county_label_lookup` used?

File: R-packages/evalcast/R/get_predictions.R [hunk 3]
```diff
@@ -88,6 +117,10 @@ get_predictions_single_date <- function(forecaster,
                       geo_type = geo_type,
                       geo_values = geo_values)
     })
+  if(!is.null(apply_corrections)){
+    corrected <- apply_corrections(df)
+    df$value = corrected$corrected
```
Comment (jacobbien) [inline R-packages/evalcast/R/get_predictions.R:122]: If a user-supplied `apply_corrections` were to reorder the rows, then this could be a problem.  Perhaps a `left_join` would be safer?

File: R-packages/evalcast/data-raw/county_label_lookup.R (hunk 2 above)
Comment (dajmcdon) [reply R-packages/evalcast/data-raw/county_label_lookup.R:2]: I'm not sure where this came from. I must have screwed up the git. I think this script and the associated data file below should just be deleted.

File: R-packages/evalcast/tests/testthat/test-caching.R (hunk 1 above)
Comment (jacobbien) [reply R-packages/evalcast/tests/testthat/test-caching.R:31]: Also, perhaps you could add a test that the predictions cards from the two runs are identical?

**Comment by jacobbien** [comment]
Thanks, @dajmcdon, this is very nice.  I have a few more comments:
1. I agree that we'd want to add the corrections within the `get_target_response` function, but can you explain what the difficulty you allude to would be with this?
2. The attributes for a predictions card should have all the information needed to reproduce the call.  I believe we should therefore add an attribute that is the `apply_corrections` function that was used with `NA` (I suppose) in the case that no corrections were used.
3. Likewise, when we add corrections to the evaluate part, we should add an attribute `apply_corrections` to the scorecard
4. Then when downstream functions that take in a list of scorecards, we will need to make sure that the same corrections were used for scorecards in the list.  The one difficulty I could foresee here is if `apply_corrections` involves randomness.  I think one should discourage that since we don't want forecasters compared with each having its own ground truth data.  Actually, the simplest thing might be to just check that the response column in each scorecard being compared is in perfect agreement.

File: R-packages/evalcast/R/get_predictions.R (hunk 3 above)
Comment (dajmcdon) [reply R-packages/evalcast/R/get_predictions.R]: I made a number of changes to address this, adding checks and a test. Currently all tests fail due to the `covidcast::name_to_abbr()` bug (see #144, fixed by @ryantibs on #146 ).

File: R-packages/evalcast/tests/testthat/test-caching.R (hunk 1 above)
Comment (dajmcdon) [reply R-packages/evalcast/tests/testthat/test-caching.R:40]: I added a test to check the prediction cards (both the attributes and the values). There's a note there about the attribute check for @bnaras and #98 . As for the timing check, I added a direct test to make sure that the caching is enabled. Waiting to push until I can try it locally.

**Comment by dajmcdon** [comment]
@jacobbien, some comments on your comments.

1. I guess I was imagining (perhaps incorrectly), that if one wanted to correct the evaluation data, the desired outcome would be the same as in the training data (or input data). For example, we use a rolling window. But based on the download scheme in `get_target_period()`, it doesn't seem like we could (since only the response over the forecast period is downloaded).  But maybe I'm wrong about how it works. We would also be unable to make corrections that depend on any signals except the response. So, it seems like there are some structural constraints on the types of corrections that can be made here that don't exist at the prediction stage. We could try to explain all the constraints to the user, or we could make more elaborate alterations. Say there's just one forecast date, download up to that date  for two `as_of`s, apply the corrections to everything and then load the memoised second `as_of` at evaluation. This all seemed fairly complicated, so I figured I'd ignore it for now. But maybe we need it?
2. I think I fixed this. But only in `get_predictions()`. I've been unable to find if they get set elsewhere checked against (in which case, I broke something).
3. The corrected prediction card attribute seems to propagate, but point 1 above.
4. Our own corrections ARE random (I abandoned the seed setting some time back). So this would pose a problem. Though, could be fixed by referencing the same corrected data for all forecasters rather than downloading separately. 

Relatedly, we have been storing the actual corrected values, not just the function to generate them. I was thinking that in our pipeline, I would make the `apply_corrections` function we pass in store the results for visualization, etc after the run. But maybe we actually want to "download and store" all the data necessary at once (though this sounds like just building upstream_df). That would allow the corrections but break lots of other things.

File: R-packages/evalcast/tests/testthat/test-caching.R (hunk 1 above)
Comment (bnaras) [reply R-packages/evalcast/tests/testthat/test-caching.R:55]: I pushed a fix to #144 so that you can try your tests.  I am not sure building the caching directly into the function is the right approach as that might screw up experiments where real-time changes may take place to the data. It seems it would be better if the user is given the option of using caching by means of a configuration setting...

Comment (bnaras) [reply R-packages/evalcast/tests/testthat/test-caching.R:55]: @jacobbien To answer your question about checking if caching occured,  if one uses the file system, there must be objects in the file system one can look at for the timestamp...

Comment (dajmcdon) [reply R-packages/evalcast/tests/testthat/test-caching.R:57]: I haven't been able to get it to work. There are two commented out tests that use `memoise::has_cache()` to actually examine the cache. One issue with the current approach (supporting @bnaras's point) is that the in-memory version is hard to access. But it greatly simplifies the package development. We could prompt on load to warn of the behavior and suggest caching locally instead. This `memoise` creates unintuitive file names though.

**Comment by dajmcdon** [comment]
I'm not sure where this got left. For the moment, we need the corrections if we're going to run our pipeline off of this.

Alternatively, I can put a fork somewhere.

**Comment by jacobbien** [comment]
1.  Thanks for this explanation.  I had forgotten that `apply_corrections()` needs to take in the history of the variable (and even other variables) in order to decide what corrections to make.  However, the good news is that if I'm not mistaken things may be easier than you are describing.  When downloading data within `evaluate_predictions()`, we don't have to worry about the "as of principle" ...our goal is to get the ground truth data as best we know it as of the current time we are running `evaluate_predictions` (not as of the forecast date).  This means that the `apply_corrections()` function could on its own download whatever it wants internally.  To be clear, this corrections function would operate by different rules from the `apply_corrections()` function we would use in the `get_predictions()` so for clarity we would want to call this one something different.  It could take in that data frame in order to know what target period dates/locations we care about and then do its own thing using all the information it wants to make the best corrections possible.

2. Thanks.  This is why we need more unit tests.  I've written some, but we should have more for cases like this.

3. Right.  Can't address this part until we decide on 1.

4.  For reproducibility it seems that setting a seed would always be desirable, but I'm probably missing something here.

Yeah, I can understand the logic of wanting to store the corrections themselves.
